packages/storage/src/userStorage.ts

```diff
@@ -623,14 +627,97 @@ export class UserStorage {
     return { ...metadata, ...getOrCreateResponse };
   }
 
-  private static async getExistingBucket(
-    client: Buckets,
-    bucketName: string,
-    threadId: string,
-  ): Promise<Root | undefined> {
-    const existingRoots = await client.existing(threadId);
+  /**
+   * shareViaPublicKey shares specified files to users who owns the specified public keys.
+   *
+   * @example
+   * ```typescript
+   * const result = await spaceStorage.shareViaPublicKey({
+   *   publicKeys: [{
+   *      id: 'user@email.com', // or any identifier for the user
+   *      pk: 'user-pk-hex-or-multibase', // optional, omit if user doesn't exist yet, it would generate temp access key
+   *   }],
+   *   paths: [{
+   *      bucket: 'personal',
+   *      path: '/file/path/here'
+   *   }],
+   * });
+   *
+   * ```
+   */
+  public async shareViaPublicKey(request: ShareViaPublicKeyRequest): Promise<ShareViaPublicKeyResponse> {
+    validateNonEmptyArray('publicKeys', request.publicKeys);
+    validateNonEmptyArray('paths', request.paths);
+    const client = this.getUserBucketsClient();
+    const userKeys = await this.normalizeShareKeys(request.publicKeys);
+    const paths = await this.normalizeFullPaths(client, request.paths);
+
+    // eslint-disable-next-line no-restricted-syntax
+    for (const userKey of userKeys) {
+      // eslint-disable-next-line no-restricted-syntax
+      for (const path of paths) {
+        const roles = new Map();
+        roles.set(userKey.pk, PathAccessRole.PATH_ACCESS_ROLE_WRITER);
+        await client.pushPathAccessRoles(path.key, path.fullPath.path, roles);
+      }
+
+      // TODO: Send a mailbox message to new user
+    }
 
-    return existingRoots.find((root) => root.name === bucketName);
+    return {
+      publicKeys: userKeys.map((keys) => {
+        return {
+          id: keys.id,
+          pk: keys.pk,
+          type: keys.type,
+          tempKey: keys.tempKey,
+        };
+      }),
+    };
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  private normalizeShareKeys(publicKeys: SharePublicKeyInput[]): Promise<SharePublicKeyOutput[]> {
```

Comment:
Why not a static func?

Reply:
True. I felt I might needs to do some manipulation/derivations later on with `this.spaceUser` for temp keys later, so I decided to leave it a class method pending review.